### DISCUSSION TITLE
Database Writer  The index 4 is out of range.

### DISCUSSION THREAD

**Discussion by ClarkeBar65** [nextgenhealthcare/connect#5751] [Q&A]
This is driving me mad.  I'm receiving the following error:
ERROR  (com.mirth.connect.connectors.jdbc.DatabaseDispatcher:114): An error occurred in channel "DAT TRACKER": Failed to write to database com.microsoft.sqlserver.jdbc.SQLServerException: The index 4 is out of range.

Super simple MS SQL table with 3 fields.  The SQL in the destination channel is:
INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB)
VALUES (${PatientName},${PatShortSSN},${PatDOB} )

I guess I've missed something but not sure what.  Thanks in advance!

**Comment by jonbartels** [comment]
Show the exact code. This kind of JDBC error usually comes up from mismatched `?` substitution tokens in paramaterized queries.

Are you building the SQL as a large string or using paramaterized queries?

Do any of your variables contain special characters?

Remember the velocity syntax only applies to XML so:
```
var bad = 'INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB)
VALUES (${PatientName},${PatShortSSN},${PatDOB} )'

//bad is literally the string "INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB) VALUES (${PatientName},${PatShortSSN},${PatDOB} )"

var good = <SQLText>INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB)
VALUES ('${PatientName}','${PatShortSSN}','${PatDOB} ')
</SQLText>

//good.toString would be "INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB)
VALUES ('Bob','1234', '1900-01-01')"
```

**Reply by ClarkeBar65** [reply]
![MirthDest](https://user-images.githubusercontent.com/130484636/231248356-65887d14-4915-41c5-af93-0354994025af.jpg)
![MirthErrorLog](https://user-images.githubusercontent.com/130484636/231248357-f79ecefc-f274-486b-ae7d-dc8679b729ad.jpg)

@jonbartels  Thanks - I've attached 2 pics ...That's all the code there is.  I'm not sure if you are looking for something different.  The variables were created from the HL7 mapping per some of the YouTube videos I watched.

**Reply by jonbartels** [reply]
Hmm that should work. For the message that failed, does the patient name, shortssn, or DOB:

* have "?" or "," characters in it?
* Have spaces?
* Is null or blank?

**Reply by ClarkeBar65** [reply]
I did write the values to the Server Log and they are definitely filled.  Perhaps its my setup - I'm running this inside a VMWare  VM for testing purposes and It has MS SSMS v18.8 installed, tho Mirth Connect does see and can read the fields to setup that insert.  Java version perhaps?  

![MirthSvrLog](https://user-images.githubusercontent.com/130484636/231255668-3a0028aa-8b72-44f9-9634-032b60bdd8a9.jpg)

**Reply by jonbartels** [reply]
No. This has to be a problem with the query construction. 

What happens if you quote your strings? 
`INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB) VALUES ('${PatientName}','${PatShortSSN}','${PatDOB}')`

**Reply by ab-20-m** [reply]
The PatientName has a "," in it. That will turn your SQL statements into roughly:
`INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB)
VALUES (JOHN,DOE,8675309,01011950)`

**Reply by ClarkeBar65** [reply]
Quoted strings still gives me "An error occurred in channel "DAT TRACKER": Failed to write to database
com.microsoft.sqlserver.jdbc.SQLServerException: The index 1 is out of range." 

I did try taking the comma out of the name as well and still got the above error.  Now you know why I'm going bonkers LOL

**Reply by ClarkeBar65** [reply]
Just for further info I tried doing this and got the error as well:

INSERT INTO MirthTest (PatientName, PatShortSSN, PatDOB)
VALUES ('DOE,JOHN','D0001','1234' )

this is what I used to create the table:
CREATE TABLE [dbo].[MirthTest](
	[PatientName] [varchar](75) NULL,
	[PatShortSSN] [varchar](10) NULL,
	[PatDOB] [varchar](16) NULL,
)
GO

**Reply by rjkroll** [reply]
Did you really create the table without any defined column lengths?   If you declare a column as a CHAR , NCHAR , VARCHAR or NVARCHAR without a length, SQL Server reads that as a length of 1.

**Reply by pacmano1** [reply]
Yeah - probably meant DEFAULT NULL also.    Postgres allows no lengith on varchar, MSSQL needs one.

**Reply by pacmano1** [reply]
And using mixed case column names is my least favorite thing ever.

**Reply by ab-20-m** [reply]
I just tested a simple Channel Reader/DB Writer with the following configuration without any issues:
Windows Server 2019
SQL Server 2019 Dev Edition
Microsoft OpenJDK 11.0.18
Mirth 4.3.0

If you look at the errored messages in the message viewer the sent should show what Mirth was trying to use for the query and parameters values.

Did you swap out the MS SQL JDBC Driver for a newer version?

**Reply by pacmano1** [reply]
![image](https://user-images.githubusercontent.com/44065187/231500050-f5e2835a-a0a6-4083-99a2-cbca33a2fcb5.png)

**Reply by ab-20-m** [reply]
I made my columns varchar(50). The MS SQL driver correctly reported the truncation error.

**Reply by pacmano1** [reply]
I just meant his table is messed up - not your stuff.  @rjkroll mentioend that.

**Reply by ClarkeBar65** [reply]
Sorry my bad I posted the wrong create.  Yes the field sizes are defined (SQL below).  These fields were pulled over from a much larger table so I could troubleshoot this issue on a smaller table.  That larger table has all the fields in camel case - not something I created or can even change at this point unfortunately.  I did notice the MS OpenJDK 11.0.18 above - I do not have that should I install that?

SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
CREATE TABLE [dbo].[MirthTest](
	[PatientName] [varchar](75) NULL,
	[PatShortSSN] [varchar](10) NULL,
	[PatDOB] [varchar](16) NULL,
)
GO

**Reply by pacmano1** [reply]
no, you need `varchar(50)` - i.e. a length for varchar

**Reply by pacmano1** [reply]
```sql
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
CREATE TABLE [dbo].[MirthTest](
[PatientName] varchar(50),
[PatShortSSN] varchar(50),
[PatDOB] varchar(50),
)
GO
```

**Reply by ClarkeBar65** [reply]
Copy/Paste hides the size.  Weird - I took out the () to show.

	[PatientName] [varchar] 75 NULL,
	[PatShortSSN] [varchar] 10 NULL,
	[PatDOB] [varchar] 16 NULL
) ON [PRIMARY]
GO

**Reply by pacmano1** [reply]
Please use code blocks.   Read https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by ClarkeBar65** [reply]
I'll remember to do that.  I recreated the table without the NULL.  Same error.  Should I install that MS JDK 11.0.18?

[2023-04-12 11:53:29.375]  ERROR  (com.mirth.connect.connectors.jdbc.DatabaseDispatcher:114): An error occurred in channel "DAT TRACKER": Failed to write to database
com.microsoft.sqlserver.jdbc.SQLServerException: The index 1 is out of range.

**Reply by pacmano1** [reply]
Please paste a _screenshot_ of the table definition from SQL Server Management Studio after it is created.

**Reply by ClarkeBar65** [reply]
Here you go.
![MirthTest](https://user-images.githubusercontent.com/130484636/231532084-6a870b30-07e0-47c8-9d9d-71d0763f8b52.jpg)

**Reply by pacmano1** [reply]
Creating the table and inserting a row, note exactly how it looks on the left:

![image](https://user-images.githubusercontent.com/44065187/231613233-8238694f-a033-4e2d-b5b0-f4bc5af8f1fb.png)

Setting some var in the connector map to be used in the next step:

![image](https://user-images.githubusercontent.com/44065187/231614910-4cc2c764-d0e2-4735-946c-31d9b68d1983.png)

The destination with the vars placed:

![image](https://user-images.githubusercontent.com/44065187/231614929-05a91178-9046-472e-838e-d76e28c2a99b.png)

The result after sending one message:

![image](https://user-images.githubusercontent.com/44065187/231614991-62b5c49f-8e5f-4751-8cff-011b7ad909c1.png)

**Reply by ClarkeBar65** [reply]
I'm not quite sure what you are asking me to do here.  Make a new step in my transformer for testing?  I already have a source transformer (HL7) that I have verified loads into the variables.

**Reply by pacmano1** [reply]
Is your stuff working?  If yes, mark the question answered.

If your stuff is not working - I posted screenshots that show that it should all work if done correctly.

**Reply by ClarkeBar65** [reply]
No I'm not working yet.  I have already previously done the DB insert thru SMS like you show.  I'm not quite sure how to setup that destination transformer like you have.  Sorry complete noob.

**Comment by ClarkeBar65** [comment]
OK I can't explain it - suddenly working.  I did make a couple of changes.  In the Destination I added logging of the 3 variables to see if they were getting loaded and I also added 'dbo.' in front of the table name.  As soon as I switched to my real table I started getting that "Index X of of range", switching back to MirthTest table resulted in the error as well.  I rebooted the VM and then it was working again - go figure.  So some weird cache issue?  

![Mirth1](https://user-images.githubusercontent.com/130484636/232820952-f382155f-
![Mirth2](https://user-images.githubusercontent.com/130484636/232821008-4609960d-c795-410b-8757-51bdb1de6b42.jpg)
c34c-4d72-9423-b7a197a884bc.jpg)
![Mirth3](https://user-images.githubusercontent.com/130484636/232821031-0163a6d1-7fff-438d-8e68-79bee53307d9.jpg)
![Mirth4](https://user-images.githubusercontent.com/130484636/232821049-0d766476-0a14-423c-b3e8-7f6882d96439.jpg)

**Comment by ClarkeBar65** [comment]
I now have it inserting into the ApplianceTrans (at least a few fields) table properly.  Apparently the fix was just to stop and start the channel after deploying it.  That seems to clear whatever the DB didn't like as far as that Index error is concerned.
